llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -1688,23 +1962,85 @@ Error MCCASBuilder::createLineSection() {
-    if (!Length)
-      return Length.takeError();
-    StringRef LineTableData = DebugLineStrRef.substr(
-        SectionStartOffset, SectionReader.getOffset() - SectionStartOffset);
-    auto DbgLineRef = DebugLineRef::create(*this, LineTableData);
-    if (!DbgLineRef)
-      return DbgLineRef.takeError();
-    addNode(*DbgLineRef);
+  DWARFDataExtractor LineTableDataReader(DebugLineStrRef,
+                                         Asm.getBackend().Endian, 8);
+  auto Prologue = parseLineTableHeaderAndSkip(LineTableDataReader);
+  if (!Prologue)
+    return Prologue.takeError();
+  SmallVector<char, 0> DistinctData;
+  // Copy line table prologue into the DistinctData buffer.
+  DistinctData.append(DebugLineStrRef.data(),
+                      DebugLineStrRef.data() + Prologue->Offset);
+  SmallVector<DebugLineRef, 0> LineTableVector;
+  SmallVector<char, 0> LineTableData;
+  uint64_t *OffsetPtr = &Prologue->Offset;
+  uint64_t End =
+      Prologue->Length + (Prologue->Format == dwarf::DWARF32 ? 4 : 8);
+  uint64_t LineTableStartOffset = Prologue->Offset;
+  while (*OffsetPtr < End) {
+    DWARFDataExtractor::Cursor Cursor(*OffsetPtr);
+    auto Opcode = LineTableDataReader.getU8(Cursor);
+    if (Opcode == 0) {
+      // Extended Opcodes always start with a zero opcode followed by
+      // a uleb128 length so you can skip ones you don't know about
+      uint64_t Len = LineTableDataReader.getULEB128(Cursor);
+      if (Len == 0)
+        return createStringError(inconvertibleErrorCode(),
+                                 "0 Length for an extended opcode is wrong");
+
+      uint8_t SubOpcode = LineTableDataReader.getU8(Cursor);
+      bool IsEndSequence = false;
+      auto Err = handleExtendedOpcodesForLineTable(
+          LineTableDataReader, Cursor, SubOpcode, Len, IsEndSequence);
+      if (Err)
+        return Err;
+      if (IsEndSequence) {
+        // The current Opcode is a DW_LNE_end_sequence. It takes no operand,
+        // create a cas block here.
+        copyData(LineTableData, DebugLineStrRef, LineTableStartOffset, Cursor);
+        auto LineTable =
+            DebugLineRef::create(*this, toStringRef(LineTableData));
+        if (!LineTable)
+          return LineTable.takeError();
+        LineTableVector.push_back(*LineTable);
+        LineTableStartOffset = Cursor.tell();
+        LineTableData.clear();
+      }
+    } else if (Opcode < Prologue->OpcodeBase) {
+      bool IsSetFile = false;
+      auto Err = handleStandardOpcodesForLineTable(LineTableDataReader, Cursor,
+                                                   Opcode, IsSetFile);
+      if (Err)
+        return Err;
+      if (IsSetFile) {
+        // The current Opcode is a DW_LNS_set_file. Store file numbers in the
+        // distinct data.
+        copyData(LineTableData, DebugLineStrRef, LineTableStartOffset, Cursor);
+        LineTableDataReader.getULEB128(Cursor);
+        // The line number doesn't dedupe, so store that in the DistinctData.
```

Comment:
Why don't they? Is this an optimization for very small functions? Would be good to comment here.

Reply:
Yeah, I'm also puzzled here